quickwit-doc-mapping/src/default_doc_mapper/default_mapper.rs

```diff
@@ -0,0 +1,378 @@
+/*
+    Quickwit
+    Copyright (C) 2021 Quickwit Inc.
+
+    Quickwit is offered under the AGPL v3.0 and as commercial software.
+    For commercial licensing, contact us at hello@quickwit.io.
+
+    AGPL:
+    This program is free software: you can redistribute it and/or modify
+    it under the terms of the GNU Affero General Public License as
+    published by the Free Software Foundation, either version 3 of the
+    License, or (at your option) any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU Affero General Public License for more details.
+
+    You should have received a copy of the GNU Affero General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+use super::{
+    field_mapping_entry::DocParsingError, resolve_field_name, FieldMappingEntry, FieldMappingType,
+};
+use crate::DocMapper;
+use anyhow::bail;
+use quickwit_proto::SearchRequest;
+use serde::{Deserialize, Serialize};
+use serde_json::{self, Value as JsonValue};
+use std::{
+    collections::{HashMap, HashSet},
+    convert::TryFrom,
+    fs::File,
+    io::BufReader,
+    path::Path,
+};
+use tantivy::{
+    query::Query,
+    schema::{Field, FieldEntry, FieldType, FieldValue, Schema, SchemaBuilder, Type, STORED},
+    Document,
+};
+
+static SOURCE_FIELD_NAME: &str = "_source";
+
+/// DefaultDocMapperBuilder is here
+/// to create a valid DefaultDocMapper.
+#[derive(Default, Serialize, Deserialize, Clone)]
+pub struct DefaultDocMapperBuilder {
+    #[serde(default)]
+    store_source: bool,
+    default_search_fields: Vec<String>,
+    field_mappings: Vec<FieldMappingEntry>,
+}
+
+impl DefaultDocMapperBuilder {
+    /// Create a new `DefaultDocMapperBuilder`
+    pub fn new() -> Self {
+        Self {
+            store_source: true,
+            default_search_fields: vec![],
+            field_mappings: vec![],
+        }
+    }
+
+    /// Build a valid `DefaultDocMapper`.
+    /// This will consume your `DefaultDocMapperBuilder`.
+    pub fn build(self) -> anyhow::Result<DefaultDocMapper> {
+        let schema = self.build_schema()?;
+        let mut default_search_fields: HashMap<String, Field> = HashMap::new();
+        for field_name in self.default_search_fields.into_iter() {
+            if default_search_fields.contains_key(&field_name) {
+                bail!("Duplicated default search field: `{}`", field_name)
+            }
+            let field = resolve_field_name(&schema, &field_name).ok_or_else(|| {
+                anyhow::anyhow!(format!("Unknown default search field: `{}`", field_name))
```

Comment:
nitpick: with_context is a tiny bit lighter for this.

```
resolve_field_name(&schema, &field_name).with_context(|| format!("Unknown default search field: `{}`", field_name))
```

Reply:
I like that too